### ISSUE TITLE
Forced Subtitles Are Not Automatically Displayed During Playback

### ISSUE THREAD

**Issue by lakerssuperman** [jellyfin-archive/jellyfin-roku-legacy#850]
Issue: Forced subtitles are not automatically displayed when starting playback of a movie and must be manually enabled.
Server version: 10.8.7
Client: 1.6.1 / whiteowl HEVC transcoding PR
Subtitles: SRT files with 'forced" subtitles

Expected behavior: forced subtitles will be automatically enabled

Notes: Forced subtitles are correctly enable when playing the same movie back through a web browser.

Screenshot shows that on playback forced subtitles are not enabled.

![IMG_20221117_194406349](https://user-images.githubusercontent.com/18699905/202591917-87ece840-a474-42b8-afa1-5443433694c5.jpg)

**Comment by sevenrats** [comment]
![image](https://user-images.githubusercontent.com/79296037/208266199-3920368e-8c73-4965-a3ae-030c08b67921.png)
What is your subtitle mode?  Can you ensure that it is set to "Default" and try to reproduce this?

**Comment by lakerssuperman** [comment]
> ![image](https://user-images.githubusercontent.com/79296037/208266199-3920368e-8c73-4965-a3ae-030c08b67921.png) What is your subtitle mode? Can you ensure that it is set to "Default" and try to reproduce this?

I have subtitles set to "Only Forced"
![Subtitles](https://user-images.githubusercontent.com/18699905/208269136-081c38e4-8efb-4493-bd9c-03087ae2aeac.png)

This is what they have always been set to.  Previously, forced would display and now they don't.  I can't set it to default and see what happens if you want.

**Comment by sevenrats** [comment]
Forced subtitles used to display on Roku?  It would help a lot to provide a version number that displays them properly and then id be happy to dig into it. im not noticing a convenient place to do it in the code.

**Comment by lakerssuperman** [comment]
It was this PR back in the spring that added the functionality as I recall.
https://github.com/jellyfin/jellyfin-roku/pull/566

I will try to go back and see if I can find a version where they were working.  I haven't been on the official app until recently because much of my collection is HEVC and until the recent release I would get unnecessary transcoding (app version 1.5 through 1.6.1).  I was running various dev versions with the HEVC fix stuff and at one point forced subtitles were working as far as I knew.  I've also upgraded my Jellyfin server over that time period as the new releases came out so I guess it's possible something changed there that affected forced subtitles.

**Comment by sevenrats** [comment]
this pr should be enough you dont need to go building every version since lord knows when.

**Comment by therealguy90** [comment]
Alright, so I just checked out the existing code in Subtitles.brs, and it seems that if the user config is set to "Any Language", subtitles will always start turned off.
There's no check to "Any Language" as the preferred language, so I modified the code, also, the existing SubtitleMode values are: "None" for None, "Default" for Default, "OnlyForced" for Only Forced, "Smart" for Smart, and "Always" for Always Play.

I haven't looked at the structure for the unstable version, so I don't really know how to fix it there and make a PR, but here's the modified function for the current release, so anyone with this issue can modify their Subtitles.brs file and fix this temporarily.

```
' Identify the default subtitle track
' if "requires_text" is true, only return a track if it is textual
'     This allows forcing text subs, since roku requires transcoding of non-text subs
' returns the server-side track index for the appriate subtitle
function defaultSubtitleTrack(sorted_subtitles, require_text = false) as integer
    if m.user.Configuration.SubtitleMode = "None"
        return -1 ' No subtitles desired: select none
    end if

    for each item in sorted_subtitles
        ' Only auto-select subtitle if language matches preference
        languageMatch = (m.user.Configuration.SubtitleLanguagePreference = item.Track.Language) or (m.user.Configuration.SubtitleLanguagePreference = "")
        ' Ensure textuality of subtitle matches preferenced passed as arg
        matchTextReq = ((require_text and item.IsTextSubtitleStream) or not require_text)
        if languageMatch and matchTextReq
            if m.user.Configuration.SubtitleMode = "Default" and (item.isForced or item.IsDefault or item.IsExternal)
                return item.Index ' Finds first forced, or default, or external subs in sorted list
            else if m.user.Configuration.SubtitleMode = "Always" and not item.IsForced
                return item.Index ' Select the first non-forced subtitle option in the sorted list
            else if m.user.Configuration.SubtitleMode = "OnlyForced" and item.IsForced
                return item.Index ' Select the first forced subtitle option in the sorted list
            else if m.user.Configuration.SubtitleMode = "Smart" and (item.isForced or item.IsDefault or item.IsExternal)
                ' Simplified "Smart" logic here mimics Default (as that is fallback behavior normally)
                ' Avoids detecting preferred audio language (as is utilized in main client)
                return item.Index
            end if
        end if
    end for
    return -1 ' Keep current default behavior of "None", if no correct subtitle is identified
end function
```

Using this, forced/default subtitles will play as intended.

**Comment by 1hitsong** [comment]
Fixed by #1604
